pkg/kubelet/dockertools/kube_docker_client.go

```diff
@@ -0,0 +1,389 @@
+	"golang.org/x/net/context"
+)
+
+// kubeDockerClient is a wrapped layer of docker client for kubelet internal use. This layer is added to:
+//	1) Redirect stream for exec and attach operations.
+//	2) Wrap the context in this layer to make the DockerInterface cleaner.
+//	3) Stabilize the DockerInterface. The engine-api is still under active development, the interface
+//	is not stabilized yet. However, the DockerInterface is used in many files in Kubernetes, we may
+//	not want to change the interface frequently. With this layer, we can port the engine api to the
+//	DockerInterface to avoid changing DockerInterface as much as possible.
+//	(See
+//	  * https://github.com/docker/engine-api/issues/89
+//	  * https://github.com/docker/engine-api/issues/137
+//	  * https://github.com/docker/engine-api/pull/140)
+// TODO(random-liu): Swith to new docker interface by refactoring the functions in the old DockerInterface
+// one by one.
+type kubeDockerClient struct {
+	client *dockerapi.Client
+}
+
+// Make sure that kubeDockerClient implemented the DockerInterface.
+var _ DockerInterface = &kubeDockerClient{}
+
+// newKubeDockerClient creates an kubeDockerClient from an existing docker client.
+func newKubeDockerClient(dockerClient *dockerapi.Client) DockerInterface {
+	return &kubeDockerClient{
+		client: dockerClient,
+	}
+}
+
+// getDefaultContext returns the default context, now the default context is
+// context.Background()
+// TODO(random-liu): Add timeout and timeout handling mechanism.
+func getDefaultContext() context.Context {
+	return context.Background()
+}
+
+// convertType converts between different types with the same json format.
+func convertType(src interface{}, dst interface{}) error {
+	data, err := json.Marshal(src)
+	if err != nil {
+		return err
+	}
+	return json.Unmarshal(data, dst)
+}
+
+// convertFilters converts filters to the filter type in engine-api.
+func convertFilters(filters map[string][]string) dockerfilters.Args {
+	args := dockerfilters.NewArgs()
+	for name, fields := range filters {
+		for _, field := range fields {
+			args.Add(name, field)
+		}
+	}
+	return args
+}
+
+// convertEnv converts data to a go-dockerclient Env
+func convertEnv(src interface{}) (*docker.Env, error) {
+	m := make(map[string]interface{})
+	if err := convertType(&src, &m); err != nil {
+		return nil, err
+	}
+	env := &docker.Env{}
+	for k, v := range m {
+		env.SetAuto(k, v)
+	}
+	return env, nil
+}
+
+func (k *kubeDockerClient) ListContainers(options docker.ListContainersOptions) ([]docker.APIContainers, error) {
+	containers, err := k.client.ContainerList(getDefaultContext(), dockertypes.ContainerListOptions{
+		Size:   options.Size,
+		All:    options.All,
+		Limit:  options.Limit,
+		Since:  options.Since,
+		Before: options.Before,
+		Filter: convertFilters(options.Filters),
+	})
+	if err != nil {
+		return nil, err
+	}
+	apiContainers := []docker.APIContainers{}
+	if err := convertType(&containers, &apiContainers); err != nil {
```

Comment:
In #23506, I just converted all the types to implement the old DockerInterface with engine-api.

Following the plan in #23563, we'll gradually refactor all the functions and types to use engine-api and get rid of go-dockerclient.

In this PR, I refactored all the container related functions and types.